modules/lib/lib-portal/src/main/resources/lib/xp/portal.ts

```diff
@@ -424,6 +424,25 @@ interface GetCurrentComponentHandler {
     execute(): object;
 }
 
+
+export type Component<
+  Config extends object = object,
+  Regions extends Record<string,Region> = Record<string,Region>
+> = {
+  config :Config
+  descriptor :string
+  path :string
+  type :'page'|'layout'|'part'
+  regions :Regions
+}
+
+export type Region<
+  Config extends object = object
+> = {
+  components :Array<Component<Config>>
```

Comment:
Same here

Reply:
Also, we use `[]` in XP instead of `Array`

Reply:
And, if I'm not mistaken, in XP we prefer `interface` instead of `type`. Again, and matter of project's code style.

Reply:
This one actually matters, interfaces can be monkeyPatched, types cannot.
So strange things can happen upstream or downstream with interfaces...

Reply:
Are you talking about the declaration merging? Like, when declaring interface with the same name, they will be combined? If it's what you are talking about, then I can assure you, using `interface` is pretty much safe here. Yes, we have globally declared interfaces, where declaration merging is intentional. 
As for the imported stuff, then merging by declaring a local interface/namespace/etc. of the same name won't work. This is actually an error and the compiler will say, that `"Import declaration conflicts with local declaration ..."`. 
Simply, this will NOT work:
```ts
import {EnonicEvent} from '/lib/xp/event';

interface EnonicEvent {
    foo: string;
}
```
But this will:
```ts
import {EnonicEvent} from '/lib/xp/event';

declare module '/lib/xp/event' {
    interface EnonicEvent {
        foo: string;
    }
}
```
In the second version there is a bit of work done, so it's hardly possible to change the interface unintentionally.

Reply:
Maybe this has been "fixed" in newer versions of TypeScript. I don't remember the details about what I read anymore, but I had some problems, so I started using type instead, and all my problems went away.

Reply:
Yeah, probably. It's easy to miss changes, unless you specifically read every release notes.
Anyway, I would rather have `interface` here for the sake of code consistency. 
P.S.: There are things to improve in our definitions (like adding `readonly` to some props), but it's just not crucial and we don't have much time for it right now. Though style discussions, suggestions and PRs are pretty much welcome (like that with namespace), because there aren't many people that will use them in Enonic.

Reply:
Readonly typically just causes a lot of headache, and since the code is synchronous, if you modify a property after it is passed to a function, it's not modified before that function has returned anyway? Or am I missing something?

Reply:
Yeah, that's right. Shouldn't probably mentioned this, because it is not related to the current PR. 
But we had a discussion with Tom, who created types for Enonic libs previously in [enonic-types](https://github.com/ItemConsulting/enonic-types), where he had a lot of `readonly` for encouraging people avoid mutable code. We decided not to this right now, but look through all the code later to see if we must really avoid mutations in some places.